src/Lean/Elab/Deriving/FromToJson.lean

```diff
@@ -11,48 +11,174 @@ import Lean.Data.Json.FromToJson
+        let a := mkIdent (← mkFreshUserName `a)
+        identNames := identNames.push a
+        ctorArgs := ctorArgs.push a
+      patterns := patterns.push (← `(@$(mkIdent ctorInfo.name):ident $ctorArgs:term*))
+      let rhs ← rhs ctorInfo identNames (if userNames.size == identNames.size then some userNames else none)
+      `(matchAltExpr| | $[$patterns:term],* => $rhs:term)
 
 def mkFromJsonInstanceHandler (declNames : Array Name) : CommandElabM Bool := do
-  if declNames.size == 1 && (← isStructure (← getEnv) declNames[0]) then
-    let cmds ← liftTermElabM none <| do
-      let ctx ← mkContext "fromJson" declNames[0]
-      let header ← mkHeader ctx ``FromJson 0 ctx.typeInfos[0]
-      let fields := getStructureFieldsFlattened (← getEnv) declNames[0] (includeSubobjectFields := false)
-      let jsonFields := fields.map (Prod.snd ∘ mkJsonField)
-      let fields := fields.map mkIdent
-      let cmd ← `(private def $(mkIdent ctx.auxFunNames[0]):ident $header.binders:explicitBinder* (j : Json)
-        : Option $(← mkInductiveApp ctx.typeInfos[0] header.argNames) := OptionM.run do
-        $[let $fields:ident ← getObjValAs? j _ $jsonFields]*
-        return { $[$fields:ident := $(id fields)]* })
-      return #[cmd] ++ (← mkInstanceCmds ctx ``FromJson declNames)
-    cmds.forM elabCommand
-    return true
+  if declNames.size == 1 then
+    if (← isStructure (← getEnv) declNames[0]) then
+      let cmds ← liftTermElabM none <| do
+        let ctx ← mkContext "fromJson" declNames[0]
+        let header ← mkHeader ctx ``FromJson 0 ctx.typeInfos[0]
+        let fields := getStructureFieldsFlattened (← getEnv) declNames[0] (includeSubobjectFields := false)
+        let jsonFields := fields.map (Prod.snd ∘ mkJsonField)
+        let fields := fields.map mkIdent
+        let cmd ← `(private def $(mkIdent ctx.auxFunNames[0]):ident $header.binders:explicitBinder* (j : Json)
+          : Option $(← mkInductiveApp ctx.typeInfos[0] header.argNames) := OptionM.run do
+          $[let $fields:ident ← getObjValAs? j _ $jsonFields]*
+          return { $[$fields:ident := $(id fields)]* })
+        return #[cmd] ++ (← mkInstanceCmds ctx ``FromJson declNames)
+      cmds.forM elabCommand
+      return true
+    else 
+      let indVal ← getConstInfoInduct declNames[0]
+      let cmds ← liftTermElabM none <| do
+        let ctx ← mkContext "fromJson" declNames[0]
+        let header ← mkHeader ctx ``FromJson 0 ctx.typeInfos[0]
+        let discrs ← mkDiscrs header indVal
+        let alts ← mkAlts indVal 
+        let matchCmd ← alts.foldrM (fun xs x => `($xs <|> $x)) (←`(none))
+        let cmd ← `(private def $(mkIdent ctx.auxFunNames[0]):ident $header.binders:explicitBinder* (json : Json)
+          : Option $(← mkInductiveApp ctx.typeInfos[0] header.argNames) :=
+            $matchCmd )
+        return #[cmd] ++ (← mkInstanceCmds ctx ``FromJson declNames)
+      cmds.forM elabCommand
+      return true
   else
     return false
-
+where
+  mkAlts
```

Comment:
Just as a note for the future, I don't think we use this style of breaking the line before any parameters anywhere else. We usually try to fill each parameter line, including the one with the function name, I believe?

Reply:
ah yes, I think I initially expected the line to be bigger, I'll join them.